FV3/atmos_cubed_sphere/tools/coarse_grained_diagnostics.F90

```diff
@@ -1,24 +1,232 @@
+    coarse_diagnostics(index)%data%var3 => Atm(tile_count)%omga(is:ie,js:je,1:npz)
+
+    index = index + 1
+    coarse_diagnostics(index)%axes = 3
+    coarse_diagnostics(index)%module_name = DYNAMICS
+    coarse_diagnostics(index)%name = 'ucomp_coarse'
+    coarse_diagnostics(index)%description = 'coarse-grained zonal wind'
+    coarse_diagnostics(index)%units = 'm/s'
+    coarse_diagnostics(index)%reduction_method = MASS_WEIGHTED
+    coarse_diagnostics(index)%data%var3 => Atm(tile_count)%ua(is:ie,js:je,1:npz)
+
+    index = index + 1
+    coarse_diagnostics(index)%axes = 3
+    coarse_diagnostics(index)%module_name = DYNAMICS
+    coarse_diagnostics(index)%name = 'vcomp_coarse'
+    coarse_diagnostics(index)%description = 'coarse-grained meridional wind'
+    coarse_diagnostics(index)%units = 'm/s'
+    coarse_diagnostics(index)%reduction_method = MASS_WEIGHTED
+    coarse_diagnostics(index)%data%var3 => Atm(tile_count)%va(is:ie,js:je,1:npz)
+
+    index = index + 1
+    coarse_diagnostics(index)%axes = 3
+    coarse_diagnostics(index)%module_name = DYNAMICS
+    coarse_diagnostics(index)%name = 'temp_coarse'
+    coarse_diagnostics(index)%description = 'coarse-grained temperature'
+    coarse_diagnostics(index)%units = 'K'
+    coarse_diagnostics(index)%reduction_method = MASS_WEIGHTED
+    coarse_diagnostics(index)%data%var3 => Atm(tile_count)%pt(is:ie,js:je,1:npz)
+
+    do t = 1, n_tracers
+      call get_tracer_names(MODEL_ATMOS, t, tracer_name, tracer_long_name, tracer_units)
+      index = index + 1
+      coarse_diagnostics(index)%axes = 3
+      coarse_diagnostics(index)%module_name = DYNAMICS
+      coarse_diagnostics(index)%name = trim(tracer_name) // '_coarse'
+      coarse_diagnostics(index)%description = 'coarse-grained ' // trim(tracer_long_name)
+      coarse_diagnostics(index)%units = tracer_units
+      coarse_diagnostics(index)%reduction_method = MASS_WEIGHTED
+      if (t .gt. n_prognostic) then
+        coarse_diagnostics(index)%data%var3 => Atm(tile_count)%qdiag(is:ie,js:je,1:npz,t)
+      else
+        coarse_diagnostics(index)%data%var3 => Atm(tile_count)%q(is:ie,js:je,1:npz,t)
+      endif
+    enddo
+
+    ! iv =-1: winds
+    ! iv = 0: positive definite scalars
+    ! iv = 1: temperature
+    do p = 1, n_pressure_levels
+      write(pressure_level_label, '(I5)') pressure_levels(p)
+
+      index = index + 1
+      coarse_diagnostics(index)%pressure_level = pressure_levels(p)
+      coarse_diagnostics(index)%axes = 2
+      coarse_diagnostics(index)%module_name = DYNAMICS
+      coarse_diagnostics(index)%name = 'u' // trim(adjustl(pressure_level_label)) // '_coarse'
+      coarse_diagnostics(index)%description = 'coarse-grained ' // trim(adjustl(pressure_level_label)) // '-mb u'
+      coarse_diagnostics(index)%units = 'm/s'
+      coarse_diagnostics(index)%reduction_method = AREA_WEIGHTED
+      coarse_diagnostics(index)%data%var3 => Atm(tile_count)%ua(is:ie,js:je,1:npz)
```

Comment:
I don't quite understand this, there is a diagnostic for each pressure level, but it's the full 3d field for the data?

Reply:
Right, yeah this is a little confusing probably.  For pressure interpolated fields the underlying source data is 3D, but the diagnostic is 2D; the pressure interpolation happens on the fly only if the diagnostic is called for in the `diag_table`.  I'm not sure if there is a cleaner way of handling this.  Maybe I'll just add a comment.